### DISCUSSION TITLE
Evolve the statevector with cusvaer

### DISCUSSION THREAD

**Discussion by hthayko** [NVIDIA/cuQuantum#160] [Q&A] [answered]
Qiskit has a very convenient method [Statevector.evolve()](https://docs.quantum.ibm.com/api/qiskit/0.19/qiskit.quantum_info.Statevector#evolve) that allows applying a single gate to a statevector. 
Is there a similar method in cusvaer (or cusvaer enabled qiskit) that accomplishes the same thing on GPUs?

**Comment by ymagchi** [comment]
Hi @hthayko,

We don't have `evolve()` method in cusvaer.
cusvaer provides one class, `cusvaer.backends.StatevectorSimulator`, which has the same interface of Qiskit Aer [StatevectorSimulator](https://qiskit.org/documentation/stubs/qiskit_aer.StatevectorSimulator.html). Using that simulator, it is possible to apply a gate and get the resulting state vector.

**Comment by hthayko** [comment]
Thanks @ymagchi,

I noticed the `cusvaer.backends.StatevectorSimulator` and have tested it - it does the job indeed. 
However if I need to repeat this procedure multiple times, the constant move from cpu to gpu adds lots of friction time.

Namely, to achieve the same thing as [Statevector.evolve()](https://docs.quantum.ibm.com/api/qiskit/0.19/qiskit.quantum_info.Statevector#evolve) I need to:
1. initialize the circuit with https://docs.nvidia.com/cuda/cuquantum/latest/appliance/cusvaer.html#set-state-simple-state.
2. add the gate to the circuit and simulate
3. get the statevector (and compute some interim stats on it, say the dot product with some fixed vector - this is why I need many .evolve() runs instead of  full simulation of all gates at once)
4. repeat

Is there a way to achieve 1..3 in gpu with cusvaer or other cuquantum methods?

**Comment by ymagchi** [comment, marked as answer]
Thanks for sharing the targeting problem.
Current cusvaer doesn't have the functionality to return the state vector directly to GPU.

If the circuit fits into single GPU, it may be possible to split the simulation into its common section and post-processing:
1. use cusvaer with the initial circuit and get the resulting state vector on host.
2. copy state vector from host to device.
3. use `cuquantum.custatevec.apply_matrix` to apply one gate, which can be computed on device. ([cuquantum.custatevec.apply_matrix](https://docs.nvidia.com/cuda/cuquantum/latest/python/api/generated/cuquantum.custatevec.apply_matrix.html#cuquantum-custatevec-apply-matrix) for reference)
4. compute dot products using the state vector on device.
5. reset the state vector on device and return to 3.

Will the above approach work in your case?

**Reply by hthayko** [reply]
Thanks @ymagchi - I tried it and it works like a charm!
Its very fast too.
Any chance there is a multi-gpu version of this? Lets say we have an 8xA100 node, e.g. single node with 8 GPUs, can we use custatevec methods to keep a large statevector (split among the 8 gpus) and apply matrices to it?

**Reply by ymagchi** [reply]
> Any chance there is a multi-gpu version of this? Lets say we have an 8xA100 node, e.g. single node with 8 GPUs, can we use custatevec methods to keep a large statevector (split among the 8 gpus) and apply matrices to it?

Currently we don't provide the multi-GPU version of gate application as single API.
Instead cuStateVec library has distributed index bit swap API, and it is possible to perform gate application on distributed state vectors by performing qubit reordering and applying gates to local index bits with single-GPU gate application API. For the details, please refer to our docs: https://docs.nvidia.com/cuda/cuquantum/latest/custatevec/distributed-index-bit-swap.html# .
